### PR TITLE
[8.0] read extraCredentials regardless of the authZ type

### PR DESCRIPTION
**Pull request by TaykYoku**
Fix https://github.com/DIRACGrid/WebAppDIRAC/issues/653.

BEGINRELEASENOTES

*Core
FIX: Process extra credentials passed in the request regardless of the type of authorization.

ENDRELEASENOTES

### CODE REVIEW

**Comment by fstagni** [comment]
FYI: there's a conflict right now.

**Comment by chrisburr** [comment]
Can you explain why this PR fixes it and what exactly the issue was as I'm struggling to figure out how it works?

File: src/DIRAC/Core/Tornado/Server/private/BaseRequestHandler.py [hunk 1]
```diff
@@ -700,15 +700,19 @@ def _gatherPeerCredentials(self, grants: list = None) -> dict:
...
             grant = grant.upper()
-            grantFunc = getattr(self, "_authz%s" % grant, None)
+            grantFunc = getattr(self, f"_authz{grant}", None)
             # pylint: disable=not-callable
-            result = grantFunc() if callable(grantFunc) else S_ERROR("%s authentication type is not supported." % grant)
-            if result["OK"]:
+            result = grantFunc() if callable(grantFunc) else S_ERROR(f"{grant} authentication type is not supported.")
+            if result["OK"] and (credDict := result["Value"]):
+                # Check if client sends extra credentials
+                if extraCred := self.get_argument("extraCredentials", None):
```
[the remark this answers is not in the document]
Comment (TaykYoku) [reply src/DIRAC/Core/Tornado/Server/private/BaseRequestHandler.py:708]: All requests pass through the base class `BaseRequestHandler`.
We are interested in the `__prepare` method that prepares the request and authenticate request.
Then `AuthManager.authQuery` authorize the request where `extraCredentials` is processed:
```
      # Check if query comes though a gateway/web server
      if self.forwardedCredentials(credDict):
          self.__authLogger.debug("Query comes from a gateway")
          self.unpackForwardedCredentials(credDict)
          return self.authQuery(methodQuery, credDict, requiredProperties)
```

`forwardedCredentials` - Check whether the credentials are being forwarded by a valid source
`unpackForwardedCredentials` - Extract the forwarded credentials

**Comment by TaykYoku** [comment]
All requests pass through the base class `BaseRequestHandler`.
We are interested in the `__prepare` method that prepares the request and authenticate request.
Then `AuthManager.authQuery` authorize the request where `extraCredentials` is processed:
```
      # Check if query comes though a gateway/web server
      if self.forwardedCredentials(credDict):
          self.__authLogger.debug("Query comes from a gateway")
          self.unpackForwardedCredentials(credDict)
          return self.authQuery(methodQuery, credDict, requiredProperties)
```

`forwardedCredentials` - Check whether the credentials are being forwarded by a valid source
`unpackForwardedCredentials` - Extract the forwarded credentials

**Comment by chrisburr** [comment]
> Can you explain why this PR fixes it and what exactly the issue was as I'm struggling to figure out how it works?

You've not answered how this fixes the problem in https://github.com/DIRACGrid/WebAppDIRAC/issues/653. At a glance it feels to me like the code will do the same thing after the code is changed.

**Comment by chrisburr** [comment]
Are you changing anything with these rebases? I tested yesterday's version of this PR and it didn't appear to fix the issue.

**Comment by TaykYoku** [comment]
so you're right, I added `ThreadConfig` to set `extraCredentials` and now I'm testing

**Comment by TaykYoku** [comment]
WebApp uses `ThreadConfig` to store user credentials and target methods run in parallel in streams. `ThreadConfig` is based on `threading.local`. That is, the values created in one thread are not available in another. That is, you need to install extra credentials in the created thread, where the target method is run, not before.

**Comment by TaykYoku** [comment]
I tested this solution on the `lbcertifdiracoauth.cern.ch` server and my tests were successful. `ExtraCredentials` are now being transferred.

**Comment by fstagni** [comment]
Is this PR still needed?

**Comment by TaykYoku** [comment]
I prefer this solution because `__executeMethod` is protected as **superprivat**, the logic is at **Core** level and can be applied by simply "flagging" `USE_EXTRA_CREDENTIALS = True`.

But since there is already an alternative solution that also solves the problem:
https://github.com/DIRACGrid/WebAppDIRAC/pull/659 and https://github.com/DIRACGrid/DIRAC/pull/6174

, this PR may be closed.

**Pull request closed without merging**